flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/ListFileSystemFiles.java

```diff
@@ -0,0 +1,153 @@
+  private final int maxListingDirectSubDirs;
+
+  public ListFileSystemFiles(
+      String taskName,
+      int taskIndex,
+      TableLoader tableLoader,
+      String location,
+      long minAgeMs,
+      boolean usePrefixListing,
+      int maxListingDepth,
+      int maxListingDirectSubDirs) {
+    Preconditions.checkNotNull(taskName, "Task name should no be null");
+    Preconditions.checkNotNull(tableLoader, "TableLoad should no be null");
+
+    this.tableLoader = tableLoader;
+    this.taskName = taskName;
+    this.taskIndex = taskIndex;
+    this.minAgeMs = minAgeMs;
+    this.location = location;
+    this.usePrefixListing = usePrefixListing;
+    this.maxListingDepth = maxListingDepth;
+    this.maxListingDirectSubDirs = maxListingDirectSubDirs;
+  }
+
+  @Override
+  public void open(OpenContext openContext) throws Exception {
+    super.open(openContext);
+    tableLoader.open();
+    Table table = tableLoader.loadTable();
+    this.io = table.io();
+    this.location = location != null ? location : table.location();
+    this.specs = table.specs();
+    this.errorCounter =
+        TableMaintenanceMetrics.groupFor(getRuntimeContext(), table.name(), taskName, taskIndex)
+            .counter(TableMaintenanceMetrics.ERROR_COUNTER);
+    this.configuration = new Configuration();
+    table.properties().forEach(configuration::set);
+  }
+
+  @Override
+  public void processElement(Trigger trigger, Context ctx, Collector<String> out) throws Exception {
+    long olderThanTimestamp = trigger.timestamp() - minAgeMs;
+    try {
+      if (usePrefixListing) {
+        Predicate<FileInfo> predicate = fileInfo -> fileInfo.createdAtMillis() < olderThanTimestamp;
+        Preconditions.checkArgument(
+            io instanceof SupportsPrefixOperations,
+            "Cannot use prefix listing with FileIO {} which does not support prefix operations.",
+            io);
+
+        FileSystemWalker.listDirRecursivelyWithFileIO(
+            (SupportsPrefixOperations) io, location, specs, predicate, out::collect);
+      } else {
+        Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
+        List<String> remainingSubDirs = Lists.newArrayList();
+        FileSystemWalker.listDirRecursivelyWithHadoop(
+            location,
+            specs,
+            predicate,
+            configuration,
+            maxListingDepth,
+            maxListingDirectSubDirs,
+            remainingSubDirs::add,
+            out::collect);
+        for (String remainingSubDir : remainingSubDirs) {
+          FileSystemWalker.listDirRecursivelyWithHadoop(
+              remainingSubDir,
+              specs,
+              predicate,
+              configuration,
+              Integer.MAX_VALUE,
+              Integer.MAX_VALUE,
+              dir -> {},
+              out::collect);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Exception listing files for {} at {}", location, ctx.timestamp(), e);
```

Comment:
This should be `warn`. The Job will not stop, but we would like to highlight this to the maintainer